crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -210,31 +210,102 @@ def nodes
       protected
 
       def upgrade_controller_nodes
+        drbd_nodes = NodeObject.find("drbd:*")
+        # FIXME: prepare for cases with no drbd out there
+        return true if drbd_nodes.empty?
+
         # TODO: find the controller node that needs to be upgraded now
         # First node to upgrade is DRBD slave
         drbd_slave = ""
+        drbd_master = ""
         NodeObject.find(
           "state:crowbar_upgrade AND (roles:database-server OR roles:rabbitmq-server)"
         ).each do |db_node|
           cmd = "LANG=C crm resource status ms-drbd-{postgresql,rabbitmq}\
           | grep \\$(hostname) | grep -q Master"
           out = db_node.run_ssh_cmd(cmd)
-          unless out[:exit_code].zero?
+          if out[:exit_code].zero?
+            drbd_master = db_node.name
+          else
             drbd_slave = db_node.name
           end
         end
-        # FIXME: prepare for cases with no drbd out there
         return false if drbd_slave.empty?
 
         node_api = Api::Node.new drbd_slave
 
-        # FIXME: save the global status information that this node is being upgraded
-        node_api.upgrade
+        save_upgrade_state("Starting the upgrade of node #{drbd_slave}")
+        return false unless node_api.upgrade
+
+        # Explicitly mark node1 as the cluster founder
+        # and adapt DRBD config to the new founder situation.
+        # This shoudl be one time action only (for each cluster).
+        unless Api::Pacemaker.set_node_as_founder drbd_slave
+          save_error_state("Changing the cluster founder to #{drbd_slave} has failed")
```

Comment:
All these `return false` and `true` are suggesting to me that it would have been better to simply `raise` a custom `UpgradeFailure` exception when something goes wrong, and `rescue` it further up the stack.  Also that would allow for saving an error and raising the exception all within a single method call, rather than having to do them separately.  Something worth considering for the future (not in this PR).